File: src/io/ziheng/array/leetcode/SquaresOfASortedArray.java
package io.ziheng.array.leetcode;

import java.util.List;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;

/**
 * LeetCode 977. Squares of a Sorted Array
 * https://leetcode.com/problems/squares-of-a-sorted-array/
 */
public class SquaresOfASortedArray {
    public int[] sortedSquares(int[] A) {
        if (A == null || A.length == 0) {
            return new int[0];
        }
        // return sortedSquaresSortListWay(A);
        return sortedSquaresSortArrayWay(A);
    }
    private int[] sortedSquaresSortArrayWay(int[] arr) {
        int[] resultArray = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            resultArray[i] = arr[i] * arr[i];
        }
        Arrays.sort(resultArray);
        return resultArray;
    }
    private int[] sortedSquaresSortListWay(int[] A) {
       List<Integer> resultList = new LinkedList<>();
        for (int n : A) {
            resultList.add(n * n);
        }
        Collections.sort(resultList);
        int[] resultArray = new int[resultList.size()];
        for (int i = 0; i < resultArray.length; i++) {
            resultArray[i] = resultList.get(i);
        }
        return resultArray;
    }
}
/* EOF */

File: src/cc/array/SquaresOfASortedArray.cc
#include <vector>
#include <algorithm>
using namespace std;
/**
 * LeetCode 977. Squares of a Sorted Array
 * https://leetcode.com/problems/squares-of-a-sorted-array/
 */
class Solution {
public:
    vector<int> sortedSquares(vector<int>& A) {
        vector<int> resultVector = vector<int>(A.size());
        for (int i = 0; i < A.size(); i++) {
            resultVector[i] = A[i] * A[i];
        }
        sort(resultVector.begin(), resultVector.end());
        return resultVector;
    }
};
int main(int argc, char const *argv[]) {
    // ...
    return 0;
}
/* EOF */